packages/SwingSet/src/vats/comms/delivery.js

```diff
@@ -48,25 +48,42 @@ export function makeDeliveryKit(state, syscall, transmit, clistKit, stateKit) {
     const slots = kdata.slots.map(slot =>
       provideLocalForKernel(slot, doNotSubscribeSet),
     );
-    return harden({ body: kdata.body, slots });
+    return harden({ ...kdata, slots });
   }
 
   // dispatch.deliver from kernel lands here (with message from local vat to
   // remote machine): translate to local, join with handleSend
-  function sendFromKernel(target, method, kargs, kresult) {
-    const result = provideLocalForKernelResult(kresult);
+  function sendFromKernel(ktarget, method, kargs, kresult) {
+    const target = provideLocalForKernel(ktarget);
     const args = mapDataFromKernel(kargs, null);
+    assert(
+      state.objectTable.has(target) || state.promiseTable.has(target),
+      X`unknown message target ${target}/${ktarget}`,
+    );
+    assert(
+      method.indexOf(':') === -1 && method.indexOf(';') === -1,
+      X`illegal method name ${method}`,
+    );
+    // TODO: if promise target not in PromiseTable: resolve result to error
```

Comment:
Huh, this case is still relevant, but I'd have to think hard about what could provoke it. It feels like something that's supposed to work, but currently won't (the pipelined message will basically get ignored), but is difficult/unlikely to get triggered.

I guess the objecttable/promisetable checks above might sometimes help to catch this. Using a non-allocating check for the target would help the most: if someone pipelines a message to the controller, the result promise will not be seen (as a `result`) by any `sendFromKernel` call, which means the result promise would never get added to `state.promiseTable`, which means the pipelined message would target a promise that isn't already in the c-list, which would be an error.

Yeah, I think this makes me more urgently interested in replacing that `provideLocalForKernel` with a non-allocating `getLocalForKernel`, which would remove the need for the objectable/promisetable check (if `provideLocalForKernel` was used, and it needed to allocate, the "naked lref" it returned would be in the c-list but not in a table, which is a weird state that we should just make impossible). Let's see if we can keep that as an invariant: it should not be possible to have an `lref` that isn't also in either `state.objectTable` or `state.promiseTable`, which means the function that allocates the lref should also immediately create the table entry.

Reply:
Note to self: the fix (to correctly handle pipelined messages) will probably involve `controller.js` adding the result promise to the meta-object table. But note that two of the methods resolve to `undefined`, so only `doAddIngress` resolves to something that could be the target of a pipelined method (which unfortunately is a juicy candidate for pipelining). The new `controller.js` needs to remember a mapping from the result promise for `doAddIngress` to the ingress `lref` that it allocated, and if a subsequent message is aimed at the old result promise, it must forward that message back into `sendFromKernel(ingressLref)`.

Reply:
Your fix sounds very wrong to me; the meta-object table is at the wrong level of abstraction for this.  (It's also not very dynamic nor is it intended to be, which it would have to be to do this.)  But in any case I don't see now the problem case you are concerned about could ever happen.  The promise table by definition contains all the promises (modulo bugs in the comms vat code which the various asserts should catch) and I don't think pipelining enters into consideration here.

Reply:
Oh I should be more specific, this is about somebody pipelining messages to the *controller* itself (the metaobject). Like if some `bootstrap.js` does:

```js
const p1 = vats.comms~.addIngress(remoteID, roid);
p1~.hello(args);
```

I think you're right that the problem I'm worried about could never happen, but I suspect it's a different problem than you were envisioning.

What I was thinking about is:

* the `addIngress` is  routed to `controller.js`, to `doAddIngress`, which does its thing right away, and then resolves the result promise to a newly-minted vref that was just added to the kernel-facing clist
  * any subsequent messages aimed at that *vref* will go to the right place, through `sendFromKernel` as usual
* but what if the kernel delivers a message to the *result promise*?
  * the comms vat is marked as "enable pipelining", so any pipelined messages (anything aimed at a promise, rather than an object) are delivered directly to the deciding vat
  * we intend this for the use of normal messages, but the enable-pipelining flag is per-vat, so it will also apply to meta-messages
* if the comms vat receives a message for this result promise:
  * the result promise isn't in a c-list, it doesn't correspond to anything in the comms promise table
  * but it also isn't in the meta-objects set
  * so the comms vat won't know what to do with it
* we could conceivably add it to the meta-object table, so that any messages aimed at this result promise would be handed to controller.js
  * and then remember it in a table that says "result promise X was resolved to kernel-facing-clist entry Y"
  * so when controller.js gets one of these lingering messages, it can re-deliver it to itself, but to the resolved target object ID, which will flow through the clists as usual

The reason that I don't think this can happen is that `doAddIngress` resolves the result promise in the very same crank. So by the time the kernel has a chance to deliver any pipelined messages, the kernel will know what the result promise has resolved to, and the kernel will deliver it to the object-id instead of the promise-id.

The originating vat (probably `bootstrap.js`) will not have received the notification of the resolution by then.. that'll still be winding its way through the run-queue. But that's ok: the kernel knows more than the sending vat, and any attempt to send to the result promise will be translated (at the moment of delivery, `kernel.js/deliverToTarget()`) into a delivery to the object ID.

Reply:
we decide to add a comment to dispatch.js to remind ourselves about the pipelined case and the fact that it won't ever happen